modules/hisilicon/hisi_src.c

```diff
@@ -0,0 +1,261 @@
+/**
+ * @file hisi_src.c  HiSilicon sound driver - recorder
+ *
+ * Copyright (C) 2022 Dmitry Ilyin
+ */
+
+#include <threads.h>
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <acodec.h>
+
+#include "hisi.h"
+
+#include <mpi_audio.h>
+#include <hi_type.h>
+
+struct ausrc_st {
+	thrd_t thread;
+	volatile bool run;
+	void *sampv;
+	size_t sampc;
+	ausrc_read_h *rh;
+	void *arg;
+	struct ausrc_prm prm;
+};
+
+static void ausrc_destructor(void *arg)
+{
+	struct ausrc_st *st = arg;
+
+	/* Wait for termination of other thread */
+	if (st->run) {
+		debug("hisi: stopping recording thread\n");
+		st->run = false;
+		thrd_join(st->thread, NULL);
+	}
+
+	int ret = HI_MPI_AI_DisableChn(0, 0);
+	if (HI_SUCCESS != ret) {
+		printf("error %d\n", ret);
+	}
+
+	ret = HI_MPI_AI_Disable(0);
+	if (HI_SUCCESS != ret) {
+		printf("error %d\n", ret);
+	}
+
+	mem_deref(st->sampv);
+}
+
+static int read_thread(void *arg)
+{
+	AEC_FRAME_S stAecFrm;
+	AUDIO_FRAME_S stFrame;
+	int ret;
+	int dev = 0;
+	int chn = 0;
+
+	struct ausrc_st *st = arg;
+
+	while (st->run) {
+		memset(&stAecFrm, 0, sizeof(AEC_FRAME_S));
+		ret = HI_MPI_AI_GetFrame(dev, chn, &stFrame, &stAecFrm, -1);
+		if (HI_SUCCESS != ret) {
+			printf("HI_MPI_AI_GetFrame(%d, %d),"
+				" failed with %#x!\n", dev, chn, ret);
+			continue;
+		}
+
+		struct auframe af;
+
+		memcpy(st->sampv, stFrame.u64VirAddr[0], stFrame.u32Len);
```

Comment:
perhaps you can use the buffer directly here, avoiding memcpy ?

use auframe and set buffer to point to stFrame.u64VirAddr[0]